Set insecure/secure ports for upgrading requests in LayoutTests

sysrqb opened the pull request:
#### b66901d110655e63b24e80f32f4b19cd69f0ab52
<pre>
Set insecure/secure ports for upgrading requests in LayoutTests
<a href="https://bugs.webkit.org/show_bug.cgi?id=267872">https://bugs.webkit.org/show_bug.cgi?id=267872</a>
<a href="https://rdar.apple.com/121388492">rdar://121388492</a>

Reviewed by Alex Christensen.

Some requests are &quot;upgradable&quot; from insecure to secure (e.g., HTTP -&gt; HTTPS) by
the engine before the resource is loaded. Currently, the only way to test this
is by using a URL with the insecure scheme but using the secure port (e.g.,
http: and port 443). This change adds the capability to test this behavior
while using URLs that are valid. A current example of this is for testing
requests where the document&apos;s CSP declared the &apos;upgrade-insecure-requests&apos;
policy. In the future, this will help with testing upgrading mixed-content.

* LayoutTests/http/tests/security/contentSecurityPolicy/upgrade-insecure-requests/iframe-upgrade.https.html:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setPortsForUpgradingInsecureSchemeForTesting):
(WebCore::Page::portsForUpgradingInsecureSchemeForTesting const):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/PageConfiguration.h:
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::upgradeInsecureRequestIfNeeded const):
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
(API::PageConfiguration::setPortsForUpgradingInsecureSchemeForTesting):
(API::PageConfiguration::portsForUpgradingInsecureSchemeForTesting const):
* Source/WebKit/UIProcess/API/C/WKPageConfigurationRef.cpp:
(WKPageConfigurationSetPortsForUpgradingInsecureSchemeForTesting):
* Source/WebKit/UIProcess/API/C/WKPageConfigurationRef.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration _portsForUpgradingInsecureSchemeForTesting]):
(-[WKWebViewConfiguration _setPortsForUpgradingInsecureSchemeForTesting:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_historyItemClient):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _setPortsForUpgradingInsecureSchemeForTesting:withSecurePort:]):
* Source/WebKitLegacy/mac/WebView/WebViewPrivate.h:
* Tools/DumpRenderTree/TestOptions.cpp:
(WTR::TestOptions::keyTypeMapping):
(WTR::TestOptions::uint16TestRunnerFeatureValue const):
* Tools/DumpRenderTree/TestOptions.h:
(WTR::TestOptions::insecureUpgradePort const):
(WTR::TestOptions::secureUpgradePort const):
* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(runTest):
* Tools/TestRunnerShared/TestFeatures.cpp:
(WTR::merge):
(WTR::operator==):
(WTR::shouldSetDefaultPortsForWTR):
(WTR::shouldSetDefaultPortsForWPT):
(WTR::hardcodedFeaturesBasedOnPathForTest):
(WTR::parseUInt16TestHeaderValue):
(WTR::parseTestHeaderFeature):
* Tools/TestRunnerShared/TestFeatures.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::generatePageConfiguration):
* Tools/WebKitTestRunner/TestOptions.cpp:
(WTR::TestOptions::defaults):
(WTR::TestOptions::keyTypeMapping):
(WTR::TestOptions::uint16TestRunnerFeatureValue const):
* Tools/WebKitTestRunner/TestOptions.h:
(WTR::TestOptions::insecureUpgradePort const):
(WTR::TestOptions::secureUpgradePort const):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::platformCreateWebView):

Canonical link: <a href="https://commits.webkit.org/273568@main">https://commits.webkit.org/273568@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/76b9e05fe241b764bc2a02d424caef53d6107259

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35774 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14717 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37920 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38500 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32193 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36992 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/17110 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11741 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30957 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36328 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12421 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31805 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10923 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10916 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31971 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39747 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32487 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32296 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36872 "Found 1 new test failure: fast/repaint/animation-after-layer-scroll.html (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/11111 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9002 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34958 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12850 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/31634 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8167 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11616 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11926 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKitLegacy/mac/WebView/WebFrame.mm:
```diff
@@ -2338,6 +2338,17 @@ - (void)_generateTestReport:(NSString *) message withGroup:(NSString *)group
         document->reportingScope().generateTestReport(message, group);
 }
 
+- (void)_setPortsForUpgradingInsecureSchemeForTesting:(uint16_t)insecureUpgradePort withSecurePort:(uint16_t)secureUpgradePort
```
This should be in WebViewPrivate.h/WebView.mm instead of WebFrame

achristensen07 commented on Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
```diff
@@ -1043,8 +1043,19 @@ void ContentSecurityPolicy::upgradeInsecureRequestIfNeeded(URL& url, InsecureReq
     }
 
+    std::optional<std::pair<uint16_t, uint16_t>> portsForUpgradingInsecureScheme;
+    if (auto* document = dynamicDowncast<Document>(m_scriptExecutionContext.get()); document && document->page())
```
This is duplicated below

sysrqb committed:
Set insecure/secure ports for upgrading requests in LayoutTests
https://bugs.webkit.org/show_bug.cgi?id=267872
rdar://121388492

Reviewed by Alex Christensen.

Some requests are "upgradable" from insecure to secure (e.g., HTTP -> HTTPS) by
the engine before the resource is loaded. Currently, the only way to test this
is by using a URL with the insecure scheme but using the secure port (e.g.,
http: and port 443). This change adds the capability to test this behavior
while using URLs that are valid. A current example of this is for testing
requests where the document's CSP declared the 'upgrade-insecure-requests'
policy. In the future, this will help with testing upgrading mixed-content.

* LayoutTests/http/tests/security/contentSecurityPolicy/upgrade-insecure-requests/iframe-upgrade.https.html:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::setPortsForUpgradingInsecureSchemeForTesting):
(WebCore::Page::portsForUpgradingInsecureSchemeForTesting const):
* Source/WebCore/page/Page.h:
* Source/WebCore/page/PageConfiguration.h:
* Source/WebCore/page/csp/ContentSecurityPolicy.cpp:
(WebCore::ContentSecurityPolicy::upgradeInsecureRequestIfNeeded const):
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/Shared/WebPageCreationParameters.serialization.in:
* Source/WebKit/UIProcess/API/APIPageConfiguration.h:
(API::PageConfiguration::setPortsForUpgradingInsecureSchemeForTesting):
(API::PageConfiguration::portsForUpgradingInsecureSchemeForTesting const):
* Source/WebKit/UIProcess/API/C/WKPageConfigurationRef.cpp:
(WKPageConfigurationSetPortsForUpgradingInsecureSchemeForTesting):
* Source/WebKit/UIProcess/API/C/WKPageConfigurationRef.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration _portsForUpgradingInsecureSchemeForTesting]):
(-[WKWebViewConfiguration _setPortsForUpgradingInsecureSchemeForTesting:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_historyItemClient):
* Source/WebKitLegacy/mac/WebView/WebView.mm:
(-[WebView _setPortsForUpgradingInsecureSchemeForTesting:withSecurePort:]):
* Source/WebKitLegacy/mac/WebView/WebViewPrivate.h:
* Tools/DumpRenderTree/TestOptions.cpp:
(WTR::TestOptions::keyTypeMapping):
(WTR::TestOptions::uint16TestRunnerFeatureValue const):
* Tools/DumpRenderTree/TestOptions.h:
(WTR::TestOptions::insecureUpgradePort const):
(WTR::TestOptions::secureUpgradePort const):
* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(runTest):
* Tools/TestRunnerShared/TestFeatures.cpp:
(WTR::merge):
(WTR::operator==):
(WTR::shouldSetDefaultPortsForWTR):
(WTR::shouldSetDefaultPortsForWPT):
(WTR::hardcodedFeaturesBasedOnPathForTest):
(WTR::parseUInt16TestHeaderValue):
(WTR::parseTestHeaderFeature):
* Tools/TestRunnerShared/TestFeatures.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::generatePageConfiguration):
* Tools/WebKitTestRunner/TestOptions.cpp:
(WTR::TestOptions::defaults):
(WTR::TestOptions::keyTypeMapping):
(WTR::TestOptions::uint16TestRunnerFeatureValue const):
* Tools/WebKitTestRunner/TestOptions.h:
(WTR::TestOptions::insecureUpgradePort const):
(WTR::TestOptions::secureUpgradePort const):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::platformCreateWebView):

Canonical link: https://commits.webkit.org/273568@main
